### PR TITLE
Support Services

### LINKED ISSUE #13: Service Support
Hello,
The cube module is already great to quickly produce dashboards ! 
But currently, it can only deal with host statistics. Is it planned to support the service statistics?
Regards,
Yannick

### PR DESCRIPTION
**Pull request by lippserd**
fixes #13

### CODE REVIEW

**Comment by lippserd** [comment]
Action links and a control to switch between the cubes are missing.

File: application/controllers/IdoController.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                'url'   => 'cube/hosts' . ($this->params->toString() === '' ? '' : '?' . $this->params->toString())
+            ))
+            ->add('cube/services', array(
+                'label' => $this->translate('Services'),
+                'url'   => 'cube/services' . ($this->params->toString() === '' ? '' : '?' . $this->params->toString())
+            ))
+            ->extend(new DashboardAction());
+    }
+
+    public function buildCube($cube)
```
Comment (lippserd) [inline application/controllers/IdoController.php]: Rename `buildCube()` to `renderCube()` and make it protected.

File: application/controllers/ServicesController.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    public function indexAction()
+    {
+        $this->addTabs()->activate('cube/services');
+        $cube = new IdoServiceStatusCube();
+
+        $this->buildCube($cube);
+    }
+
+    public function detailsAction()
```
Comment (lippserd) [inline application/controllers/ServicesController.php]: Please remove this method.

File: application/controllers/IdoController.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if ($showSettings) {
+            $this->view->form = $this->loadForm('Dimensions')
+                ->setCube($cube)
+                ->handleRequest();
+        } else {
+            $this->setAutorefreshInterval(15);
+        }
+    }
+
+    public function detailsAction()
```
Comment (lippserd) [inline application/controllers/IdoController.php]: Move the `detailsAction()` to the module's base controller class and rely on the protected member variable `$cube`.

File: application/controllers/ServicesController.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Icinga\Module\Cube\Controllers;
+
+use Icinga\Module\Cube\Ido\IdoServiceStatusCube;
+use Icinga\Web\UrlParams;
+
+class ServicesController extends IdoController
+{
+    /** @var UrlParams */
+    protected $params;
```
Comment (lippserd) [inline application/controllers/ServicesController.php]: Please remove this variable.

File: application/controllers/IndexController.php [hunk 5]
```diff
@@ -1,89 +1,13 @@
 <?php
-// Icinga Web 2 Cube Module | (c) 2016 Icinga GmbH | GPLv2
+// Icinga Web 2 Cube Module | (c) 2019 Icinga GmbH | GPLv2
```
Comment (lippserd) [inline application/controllers/IndexController.php]: Don't change the copyright year here please.

File: application/controllers/IndexController.php [hunk 6]
```diff
@@ -1,89 +1,13 @@
...
-        foreach (array('renderLayout', 'showFullscreen', 'showCompact', 'view') as $p) {
-            $params->shift($p);
-        }
-
-        foreach ($this->params->toArray() as $param) {
-            $cube->slice($param[0], urldecode($param[1]));
-        }
-
-        return $cube;
+    public function indexAction() {
```
Comment (lippserd) [inline application/controllers/IndexController.php]: There is a newline missing.

File: library/Cube/Web/Controller.php [hunk 7]
```diff
@@ -8,11 +8,99 @@
...
-class Controller extends WebController
+abstract class Controller extends WebController
 {
     /** @var View This helps IDEs to understand that this is not ZF view */
     public $view;
 
+    /** @var \Icinga\Module\Cube\Cube */
+    protected $cube;
+
+    abstract protected function getCube();
```
Comment (lippserd) [inline library/Cube/Web/Controller.php]: Please add function docs.

File: application/controllers/IndexController.php (hunk 6 above)
Comment (theFeu) [reply application/controllers/IndexController.php]: A newline before the action?

Comment (lippserd) [reply application/controllers/IndexController.php]: A newline before the `{`.

Comment (theFeu) [reply application/controllers/IndexController.php]: 🤦‍♂️

File: library/Cube/Web/Controller.php (hunk 7 above)
Comment (lippserd) [reply library/Cube/Web/Controller.php]: It is either **return** or **get**. We don't add the **s**.

**Comment by theFeu** [comment]
@lippserd ready to merge?

**Comment by lippserd** [comment]
There are WIP commits. Please amend/squash.

**Pull request merged**